### ISSUE TITLE
Issues with casting of floating-point numbers

### ISSUE THREAD

**Issue by shmarovfedor** [esbmc/esbmc#877]
The following code lists several scenarios where ESBMC crashes (`Assertion failed: (t->sort->get_data_width() == f->sort->get_data_width()), function mk_ite, file boolector_conv.cpp, line 628.`) when using Boolector as the SMT solver. However, the same program works OK with Z3. It seems that this issue with Boolector appears when you cast `c` to anything larger/small than the first cast that has been used to initialise it in `line 2`. 
```
  1 int func() {
  2   float b, c = (int)b;
  3   //assert((int)c > 0); // Works with Boolector                                                                                                                         
  4   //assert((unsigned int) c > 0); // Works with Boolector
  5   //assert((long)c > 0); // Crashes with Boolector 
  6   //assert((short)c > 0); // Crashes with Boolector 
  7   //assert((char)c > 0); // Crashes with Boolector 
  8   //assert((float)c > 0); // Works with Boolector
  9 }
```
Further investigation of the issue revealed that there are differences in the SMT formula generation for floating-point types depending on the chosen solver. Namely, it appears that this particular problem comes from the `smt_astt smt_convt::convert_typecast_from_fpbv(const typecast2t &cast)` method. It uses `mk_smt_typecast_from_fpbv_to_sbv(smt_astt from, std::size_t width)`  which has an implementation in `z3_conv` using the available Z3 API functions, but the Boolector API does not provide support for it. As the result, ESBMC uses the implementation of this method from the `solvers/smt/fp/fp_conv.cpp` module when Boolector is used. The implementation of `mk_smt_typecast_from_fpbv_to_sbv(smt_astt from, std::size_t width)` in `solvers/smt/fp/fp_conv.cpp` has a single call to `smt_astt fp_convt::mk_to_bv(smt_astt x, bool is_signed, std::size_t width)` which seems to be quite involved (about 150 lines of code, and it might be causing the described issue), and it looks like that it is responsible for the conversion of floating point types into bit vector types.

**Comment by rafaelsamenezes** [comment]
@mikhailramalho Do you have any suggestions?

**Comment by mikhailramalho** [comment]
I'll take a look later today.

**Comment by mikhailramalho** [comment]
Fix in PR #878
